denonavr/api.py

```diff
@@ -309,3 +314,148 @@ def prepare_appcommand_body(cmd_list: Tuple[AppCommandCmd]) -> bytes:
+            try:
+                chunk = await asyncio.wait_for(
+                    self._socket_reader.read(_SOCKET_READ_SIZE), self.timeout
+                )
+            except asyncio.exceptions.TimeoutError:
+                _LOGGER.debug("Lost connection to receiver, reconnecting")
+                await self.async_disconnect()
+                await self.async_connect()
+            # pylint: disable=consider-using-enumerate
+            for i in range(0, len(chunk)):
+                # Messages are CR terminated
+                if chunk[i] != 13:
+                    data += chunk[i].to_bytes(1, byteorder='big')
+                else:
+                    await self._process_event(str(data, 'utf-8'))
+                    data = bytearray()
+
+    def register_callback(
+        self,
+        event: str,
+        callback: Callable[[str, str, str], Awaitable[None]]
+    ):
+        """Register a callback handler for an event type."""
+        # Validate the passed in type
+        if event != "ALL" and event not in TELNET_EVENTS:
+            raise ValueError("{} is not a valid callback type.".format(event))
+
+        if event not in self._callbacks.keys():
+            self._callbacks[event] = []
+        self._callbacks[event].append(callback)
+
+    def unregister_callback(
+        self,
+        event: str,
+        callback: Callable[[str, str, str], Awaitable[None]]
+    ):
+        """Unregister a callback handler for an event type."""
+        if event not in self._callbacks.keys():
+            return
+        self._callbacks[event].remove(callback)
+
+    async def _process_event(self, message: str):
+        """Process a realtime event."""
+        if len(message) < 3:
+            return None
+
+        # Event is 2 characters
+        event = message[0:2]
+        # Parameter is the remaining characters
+        parameter = message[2:]
+
+        if event == 'MV':
+            # This seems undocumented by Denon and appears to basically be a
+            # noop that goes along with volume changes. This is here to prevent
+            # duplicate callback calls.
+            if parameter[0:3] == "MAX":
+                return
+            await self._run_callbacks("MV", MAIN_ZONE, parameter)
+        elif event in ('Z2', 'Z3'):
+            if event == 'Z2':
+                zone = ZONE2
+            else:
+                zone = ZONE3
+            if parameter in ('ON', 'OFF'):
+                await self._run_callbacks("PW", zone, parameter)
+            elif parameter in ('MUON', 'MUOFF'):
+                await self._run_callbacks("MU", zone, parameter)
+            elif parameter in TELNET_SOURCES:
+                await self._run_callbacks("SI", zone, parameter)
+            elif parameter.isdigit():
+                await self._run_callbacks("MV", zone, parameter)
+        elif event in TELNET_EVENTS:
+            await self._run_callbacks(event, MAIN_ZONE, parameter)
```

Comment:
```suggestion
    async def _process_event(self, message: str):
        """Process a realtime event."""
        if len(message) < 3:
            return None
        zone = MAIN_ZONE

        # Event is 2 characters
        event = message[0:2]
        # Parameter is the remaining characters
        parameter = message[2:]
        
        if event == "MV":
            # This seems undocumented by Denon and appears to basically be a
            # noop that goes along with volume changes. This is here to prevent
            # duplicate callback calls.
            if parameter[0:3] == "MAX":
                return

        if event in ("Z2", "Z3"):
            if event == "Z2":
                zone = ZONE2
            else:
                zone = ZONE3

            if parameter in ("ON", "OFF"):
                event = "PW"
            elif parameter in TELNET_SOURCES:
                event = "SI"
            elif parameter.isdigit():
                event = "MV"
            elif parameter[0:2] in TELNET_EVENTS:
                event = parameter[0:2]
                parameter = parameter[2:]

        if event not in TELNET_EVENTS:
            return

        await self._run_callbacks(event, zone, parameter)
```
Maybe something like this to have less `self._run_callbacks` calls.
Additionally, for zones most of the events could be just shifted by 2 letters. This would cover some more commands of the zones which seem to exist according to [this document](https://www.heimkinoraum.de/upload/files/product/IP_Protocol_AVR-Xx100.pdf).